caffe2/core/storage.h

```diff
@@ -32,6 +35,35 @@ class StorageImpl {
   explicit StorageImpl(DeviceType device_type) : device_type_(device_type) {}
   StorageImpl(DeviceType device_type, TypeMeta data_type)
       : data_type_(data_type), device_type_(device_type) {}
+  template <typename Deleter = MemoryDeleter>
+  StorageImpl(
+      DeviceType device_type,
+      TypeMeta data_type,
+      void* src,
+      TIndex size,
+      size_t capacity,
+      Deleter d = nullptr)
+      : data_type_(data_type), device_type_(device_type) {
+    CAFFE_ENFORCE_WITH_CALLER(
+        data_type_.id() != TypeIdentifier::uninitialized(),
+        "To create storage with a raw external pointer you need to have meta "
+        "already initialized.");
+    // Check if the deleter is a MemoryDeleter and is a simple nullptr.
+    if (std::is_same<MemoryDeleter, Deleter>::value &&
+        reinterpret_cast<MemoryDeleter*>(&d)[0] == nullptr) {
```

Comment:
I think this code will trip some strict aliasing Werror builds internally? (May be safer to cast to `void*` first, and then to `MemoryDeleter*`)